docs/component-api.md

```diff
@@ -0,0 +1,112 @@
+
+# Component API definition and use
+
+We have chosen as Nunjucks as the templating language for GOV.UK Frontend components. We expose those templates as reusable chunks of code: macros. Developers import macros into their application, call them as per documentation and provide data to its parameters.
+
+To provide a level of consistency for developers we have standardised parameter names, their expected input, use and placement. There are expectations, and  if so they are documented accordingly.
+
+## Specifying content
+When providing *content* to a macro, say for a label or a button, we accept two parameter options:
+
+ - `text` accepts a plain string and is the default way of passing content
+ - `html` accepts html markup. In the template we will not escape html so it will be rendered. In a scenario where both text and html are set, html parameter will take precedence over text.
+
+Example:
+
+`{{ govukButton({"text": "Button text"}) }}`
+
+`{{ govukButton({"html": "Button <span class='bold'>text</span>"}) }}`
+
+Example of implementing logic in a component template:
+
+`{{ params.html | safe if params.html else params.text }}`
+
+Example shows that if `html` and `text` parameters are present, then `html` takes precedence over `text` and we are not escaping it.
+
+## Naming attributes
+
+We should use **camelCase** for naming attributes.
+
+If a component depends on another component, we group the attributes for the dependent component inside an object, where the name of the object is the name of the component using **camelCase** convention. In case of ambiguity we prefix the component name.
+
+Example of a component depending on another component
+```
+{{ govukLabel({
+	"text": "Label text",
+	"errorMessage": {
+		"text": "Error message"
+	}
+}) }}
+```
+
+Example of a component depending on two other components
+```
+{{ govukInput({
+	"name": "example-input",
+	"label": {
+		"text": "Label text"
+	},
+	"errorMessage": {
+		"text": "Error message"
+	}
+}) }}
+```
+
+## Mimic HTML attribute names
+When there is a need to specify html attributes, such as *checked, disabled, id, name*, etc, and they map directly, we use the same parameter name. We use boolean value to check and render the attribute.
+
+Example:
+
+`{{ govukButton({"disabled": true}) }}`
+
+`{{ govukCheckbox({"checked": true}) }}`
+
+
+## Defining additional HTML attributes
```

Comment:
Should we drop 'additional' here - don't these override any attributes set elsewhere?

Reply:
No, they'll be appended blindly to the end of the tag. If you used the `classes` argument and then provided `classes` in `attributes` as well then you'd end up with two `class` attributes on your tag.

Reply:
hopefully no one will add "class" in there, as we're making it clear what "classes" is for in macro arguments table, but :man_shrugging:

Reply:
I just tried adding two class definitions to an element

`<div class="a" class="b">`

Chrome only used the first class and ignored the second. Is that what we're expecting from the `attributes` feature?

Reply:
yes, safari, chrome and firefox just accept first class definition, which is great and works in a way that makes "atributes" work, even if they add "class" in there

Reply:
I'm just wondering if we should add attributes to the start of the tag instead of the end, so custom attributes always apply?

Reply:
I'd be inclined to say to, because then they can override our specific class, which i don't think we want. All other additional attributes will be applied.

Reply:
We can't do anything too clever here because of the limitations of nunjucks. Do we think there's a simple change we can make to the documentation to make this clearer?

Reply:
ok yeh happy to see how this goes, I'm just wondering if there'll be any situation where someone using components really needs to override something we've done, in which case it might be better to add attributes to the start of the tag, rather than the end. But I can see that's also dangerous and can break things so lets see how it goes with current method.

Reply:
Could/should we create an explicit attribute for doing that? 'classOverride = true' kind of thing?

Reply:
thats true, we could investigate that if there's a need